server/src-lib/Hasura/Metadata/Class.hs

```diff
@@ -0,0 +1,135 @@
+
+  -- Scheduled triggers
+  -- TODO:-
+  -- Ideally we would've liked to avoid having functions that are specific to
+  -- scheduled/cron triggers and instead have functions that provide a generic
+  -- 'event storage and retrieval' interface but we'll have to change a lot of
+  -- existing code for scheduled and cron triggers. We can get to this after the
+  -- multi-source work is done. See Note [Todo: Common interface for eventing sub-system]
+  getDeprivedCronTriggerStats :: m [CronTriggerStats]
+  getScheduledEventsForDelivery :: m ([CronEvent], [OneOffScheduledEvent])
+  insertScheduledEvent :: ScheduledEventSeed -> m ()
+  insertScheduledEventInvocation :: Invocation 'ScheduledType -> ScheduledEventType -> m ()
+  setScheduledEventOp :: ScheduledEventId -> ScheduledEventOp -> ScheduledEventType -> m ()
+  unlockScheduledEvents :: ScheduledEventType -> [ScheduledEventId] -> m Int
+  unlockAllLockedScheduledEvents :: m ()
+
+instance (MonadMetadataStorage m) => MonadMetadataStorage (ReaderT r m) where
+  getDeprivedCronTriggerStats        = lift getDeprivedCronTriggerStats
+  getScheduledEventsForDelivery      = lift getScheduledEventsForDelivery
+  insertScheduledEvent               = lift . insertScheduledEvent
+  insertScheduledEventInvocation a b = lift $ insertScheduledEventInvocation a b
+  setScheduledEventOp a b c          = lift $ setScheduledEventOp a b c
+  unlockScheduledEvents a b          = lift $ unlockScheduledEvents a b
+  unlockAllLockedScheduledEvents     = lift unlockAllLockedScheduledEvents
+
+instance (MonadMetadataStorage m) => MonadMetadataStorage (Tracing.TraceT m) where
+  getDeprivedCronTriggerStats        = lift getDeprivedCronTriggerStats
+  getScheduledEventsForDelivery      = lift getScheduledEventsForDelivery
+  insertScheduledEvent               = lift . insertScheduledEvent
+  insertScheduledEventInvocation a b = lift $ insertScheduledEventInvocation a b
+  setScheduledEventOp a b c          = lift $ setScheduledEventOp a b c
+  unlockScheduledEvents a b          = lift $ unlockScheduledEvents a b
+  unlockAllLockedScheduledEvents     = lift unlockAllLockedScheduledEvents
+
+{- Note [MonadMetadataStorage class constraint]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+All functions in @'MonadMetadataStorage' type class are expected to raise a
+@'QErr' exception. These exceptions are caught at the call site and not expected
+to be embeded in the underlying function's monad. To acheive this, we take help
+of @'MetadataStorageT' transformer which is a newtype wrapper over @'ExceptT QErr'.
+And do following:-
+
+1. Functions which implements @'MonadMetadataStorage' have
+   `MonadMetadataStorage (MetadataStorageT m)` as type class constraint.
+   Hence, we derive instance for any monad, let's say 'MyMonad', as
+
+   `instance MonadMetadataStorage (MetadataStorageT MyMonad) where`
+
+2. We use @'runMetadataStorageT' to catch exception thrown by the class methods
+   at call site.
+
+-}
```

Comment:
Here’s my suggested rewording of this comment, as promised. I’ve changed the name of the `Note` to be a little more descriptive as to what’s actually discussed here, but I haven’t renamed the various references (since this is just a GH PR suggestion). If you accept this suggestion, please rename the references, too!

```suggestion
{- Note [Generic MetadataStorageT transformer]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
All methods of the MonadMetadataStorage class may fail, which we represent in
the usual way using a MonadError superclass:

    class MonadError QErr m => MonadMetadataStorage m

However, unusually, the location where we pick a concrete MonadMetadataStorage
instance is not a context where we can handle errors, and as such the monad at
that point has no MonadError instance! Instead, clients of MonadMetadataStorage
are expected to handle errors /locally/, even though the code is parameterized
over an arbitrary metadata storage mechanism.

To encode this, we take a slightly unorthodox approach involving the auxiliary
MetadataStorageT transformer, which is really just a wrapper around ExceptT:

    newtype MetadataStorageT m a
      = MetadataStorageT { unMetadataStorageT :: ExceptT QErr m a }

We then define MonadMetadataStorage instances on a transformer stack comprising
both MetadataStorageT and a concrete base monad:

    instance MonadMetadataStorage (MetadataStorageT PGMetadataStorageApp)

This looks unconventional, but it allows polymorphic code to be parameterized
over the metadata storage implementation while still handling errors locally.
Such functions include a constraint of the form

    MonadMetadataStorage (MetadataStorageT m) => ...

and use runMetadataStorageT at the location where errors should be handled, e.g.:

    result <- runMetadataStorageT do
      {- ... some metadata operations ... -}
    case result of
      Left err -> ...
      Right value -> ...

In other words, runMetadataStorageT serves as a marker that says “I’m going to
handle exceptions raised by metadata operations right here,” which allows them
to be handled more locally than the point at which the concrete
MonadMetadataStorage instance (and thus the particular metadata storage
implementation) is actually chosen. -}
```